hail/src/main/scala/is/hail/backend/service/Worker.scala

```diff
@@ -44,21 +69,114 @@ class WorkerTimer() {
 }
 
 object Worker {
-  private val log = Logger.getLogger(getClass.getName())
+  val myRevision = HAIL_REVISION
 
-  def main(args: Array[String]): Unit = {
-    if (args.length != 2) {
-      throw new IllegalArgumentException(s"expected two arguments, not: ${ args.length }")
+  private[this] val log = Logger.getLogger(getClass.getName())
+  private[this] val revisions = mutable.Map[String, Array[String] => Unit]()
+  private[this] val scratchDir = {
+    val x = System.getenv("HAIL_WORKER_SCRATCH_DIR")
+    if (x == null) "" else x
+  }
+  private[this] val workerGCSFS = retryTransientErrors {
+    using(new FileInputStream(s"/worker-key.json")) { is =>
+      new GoogleStorageFS(IOUtils.toString(is, Charset.defaultCharset().toString()))
+    }
+  }
+
+  private[this] def withThreadContextClassLoader[T](custom: ClassLoader)(f: => T): T = {
+    val original = Thread.currentThread().getContextClassLoader()
+    try {
+      Thread.currentThread().setContextClassLoader(custom)
+      f
+    } finally {
+      Thread.currentThread().setContextClassLoader(original)
+    }
+  }
+
+  private[this] val STRING_ARRAY_CLASS = Class.forName("[Ljava.lang.String;")
+  private[this] def loadMainFromLocalJar(revision: String, localJarURL: URL): Array[String] => Unit = {
+    val classLoader = new LoadSelfFirstURLClassLoader(Array(localJarURL))
+    val reflectedMain = withThreadContextClassLoader(classLoader) {
+      val workerClass = classLoader.loadClass("is.hail.backend.service.Worker")
+
+      val actualRevision = workerClass.getMethod("myRevision").invoke(null).asInstanceOf[String]
```

Comment:
Should this not be `getValue`? Or are all values thunks? I don't know how Scala reflection works.

Reply:
So, this is probably a bit not kosher. I'm using Java reflection, not Scala reflection. I spent maybe an hour fiddling with Scala reflection before I gave up. Scala compiles non `private[this]` fields as nullary methods, so I call that method. I use `null` as the `this` object because I know this is a static method (it's a val of an object).

I could try again with Scala reflection, that'd probably be more safe.